src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -70,22 +71,35 @@ public JsonClassInfo ElementClassInfo
 
         public void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Check if we are trying to build the sorted cache.
-            if (frame.PropertyRefCache == null)
-            {
-                return;
-            }
+            Debug.Assert(frame.PropertyRefCache != null);
+
+            // frame.PropertyRefCache is only read\written by a single thread -- the thread performing
+            // the deserialization for a given object instance.
+
+            List<PropertyRef> listToAppend = frame.PropertyRefCache;
+
+            // _propertyRefsSorted can be accessed by multiple threads, so replace the reference when
+            // appending to it. No lock() is necessary.
 
-            List<PropertyRef> newList;
             if (_propertyRefsSorted != null)
             {
-                newList = new List<PropertyRef>(_propertyRefsSorted);
-                newList.AddRange(frame.PropertyRefCache);
-                _propertyRefsSorted = newList.ToArray();
+                List<PropertyRef> replacementList = new List<PropertyRef>(_propertyRefsSorted);
+                Debug.Assert(replacementList.Count <= PropertyNameCountCacheThreshold);
+
+                // Verify replacementList will not become too large.
+                while (replacementList.Count + listToAppend.Count > PropertyNameCountCacheThreshold)
+                {
+                    // This code path is rare; keep it simple by using RemoveAt() instead of RemoveRange() which requires calculating index\count.
+                    listToAppend.RemoveAt(listToAppend.Count - 1);
+                }
+
+                // Add the new items; duplicates are possible but that is tolerated during property lookup.
+                replacementList.AddRange(listToAppend);
+                _propertyRefsSorted = replacementList.ToArray();
             }
             else
             {
-                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
+                _propertyRefsSorted = listToAppend.ToArray();
```

Comment:
There's a race condition here where two threads could end up creating a list, store them into _propertyRefsSorted, but obviously _propertyRefsSorted will only contain the data from one list, which means the other thread then won't see its own data when it subsequently reads from the field.  From your comments it sounds like that's ok?

Reply:
Yes that is fine and expected. This cache is only for performance and may not contain all properties -- if not found in this cache, the code that accesses this cache then falls back to the "real" metadata-based cache that contains all properties.